substrate/frame/identity/src/lib.rs

```diff
@@ -886,6 +1012,185 @@ pub mod pallet {
+		///
+		/// The authority can grant up to `allocation` usernames. To top up their allocation, they
+		/// should just issue (or request via governance) a new `add_username_authority` call.
+		#[pallet::call_index(15)]
+		#[pallet::weight(T::WeightInfo::add_username_authority())]
+		pub fn add_username_authority(
+			origin: OriginFor<T>,
+			authority: AccountIdLookupOf<T>,
+			suffix: Vec<u8>,
+			allocation: u32,
+		) -> DispatchResult {
+			T::UsernameAuthorityOrigin::ensure_origin(origin)?;
+			let authority = T::Lookup::lookup(authority)?;
+			// We don't need to check the length because it gets checked when casting into a
+			// `BoundedVec`.
+			Self::validate_username(&suffix, None).map_err(|_| Error::<T>::InvalidSuffix)?;
+			let suffix = Suffix::<T>::try_from(suffix).map_err(|_| Error::<T>::InvalidSuffix)?;
+			// The authority may already exist, but we don't need to check. They might be changing
+			// their suffix or adding allocation, so we just want to overwrite whatever was there.
+			UsernameAuthorities::<T>::insert(
+				&authority,
+				AuthorityPropertiesOf::<T> { suffix, allocation },
+			);
+			Self::deposit_event(Event::AuthorityAdded { authority });
+			Ok(())
+		}
+
+		/// Remove `authority` from the username authorities.
+		#[pallet::call_index(16)]
+		#[pallet::weight(T::WeightInfo::remove_username_authority())]
+		pub fn remove_username_authority(
+			origin: OriginFor<T>,
+			authority: AccountIdLookupOf<T>,
+		) -> DispatchResult {
+			T::UsernameAuthorityOrigin::ensure_origin(origin)?;
+			let authority = T::Lookup::lookup(authority)?;
+			UsernameAuthorities::<T>::take(&authority).ok_or(Error::<T>::NotUsernameAuthority)?;
+			Self::deposit_event(Event::AuthorityRemoved { authority });
+			Ok(())
+		}
+
+		/// Set the username for `who`. Must be called by a username authority.
+		///
+		/// The authority must have an `allocation`. Users can either pre-sign their usernames or
+		/// approve later.
+		///
+		/// Usernames must:
+		///   - Only contain lowercase ASCII characters or digits.
+		///   - When combined with the suffix of the issuing authority be _less than_ the
+		///     `MaxUsernameLength`.
+		#[pallet::call_index(17)]
+		#[pallet::weight(T::WeightInfo::set_username_for())]
+		pub fn set_username_for(
+			origin: OriginFor<T>,
+			who: AccountIdLookupOf<T>,
+			username: Vec<u8>,
+			signature: Option<T::OffchainSignature>,
+		) -> DispatchResult {
+			// Ensure origin is a Username Authority and has an allocation. Decrement their
+			// allocation by one.
+			let sender = ensure_signed(origin)?;
+			let suffix = UsernameAuthorities::<T>::try_mutate(
+				&sender,
+				|maybe_authority| -> Result<Suffix<T>, DispatchError> {
+					let properties =
+						maybe_authority.as_mut().ok_or(Error::<T>::NotUsernameAuthority)?;
+					ensure!(properties.allocation > 0, Error::<T>::NoAllocation);
+					properties.allocation.saturating_dec();
```

Comment:
Should this be done at the point just before the username is inserted/queued, rather than here? There are a few different ways it could fail between here and insertion.

I suppose it just depends whether the allocation limit is intended as pressure for registrars to make fewer mistakes to limit pointless work being done, vs purely as a storage limit

Reply:
The whole transaction should be atomic though.

Reply:
Yes and there no re-entrancy possible here either.

Reply:
Ah I didn't realise that